extensions/interactions/PencilCodeEditor/directives/oppia-interactive-pencil-code-editor.component.ts

```diff
@@ -0,0 +1,205 @@
+    });
+  }
+
+  getNormalizedCode(): string {
+    // Converts tabs to spaces.
+    return this.pce.getCode().replace(/\t/g, '  ');
+  }
+
+  ngOnInit(): void {
+    this.directiveSubscriptions.add(
+      this.playerPositionService.onNewCardAvailable.subscribe(
+        () => {
+          this.interactionIsActive = false;
+          this.pce.hideMiddleButton();
+          this.pce.hideToggleButton();
+          this.pce.setReadOnly();
+        }
+      )
+    );
+
+    this.iframeDiv = this.elementRef.nativeElement.querySelectorAll(
+      '.pencil-code-editor-iframe')[0];
+    this.pce = new PencilCodeEmbed(this.iframeDiv);
+    this.interactionIsActive = (this.lastAnswer === null);
+
+    const { initialCode } = (
+      this.interactionAttributesExtractorService.getValuesFromAttributes(
+        'PencilCodeEditor',
+        this._getAttributes()
+      ) as PencilCodeEditorCustomizationArgs);
+    this.someInitialCode = this.interactionIsActive ?
+      initialCode.value :
+      this.lastAnswer.code;
+
+    this.pce.beginLoad(this.someInitialCode);
+    this.pce.on('load', () => {
+      // Hides the error console at the bottom right, and prevents it
+      // from showing up even if the code has an error. Also, hides the
+      // turtle, and redefines say() to also write the text on the
+      // screen.
+      this.pce.setupScript([{
+        code: [].join('\n'),
+        type: 'text/javascript'
+      }]);
+
+      this.pce.showEditor();
+      this.pce.hideToggleButton();
+      if (this.interactionIsActive) {
+        this.pce.setEditable();
+      } else {
+        this.pce.hideMiddleButton();
+        this.pce.setReadOnly();
+      }
+
+      // Pencil Code automatically takes the focus on load, so we clear
+      // it.
+      this.focusManagerService.clearFocus();
+    });
+
+    let errorIsHappening = false;
+    let hasSubmittedAnswer = false;
+
+    this.pce.on('startExecute', () => {
+      hasSubmittedAnswer = false;
+    });
+
+    this.pce.on('execute', () => {
+      if (errorIsHappening || hasSubmittedAnswer) {
+        return;
+      }
+      // The first argument in the method below gets executed in the
+      // pencilcode output-frame iframe context. The code input by the
+      // user is sanitized by pencilcode so there is no security
+      // issue in this case.
+      this.pce.eval(
+        'document.body.innerHTML', // disable-bad-pattern-check
+        (pencilCodeHtml) => {
+          let normalizedCode = this.getNormalizedCode();
+
+          // Get all the divs, and extract their textual content.
+          let output = $.map(
+            $(pencilCodeHtml).filter('div'), (elem) => {
+              return $(elem).text();
+            }).join('\n');
+
+          hasSubmittedAnswer = true;
+          this.currentInteractionService.onSubmit({
+            code: normalizedCode,
+            output: output || '',
+            evaluation: '',
+            error: ''
+          } as unknown as string,
```

Comment:
this object is clearly not a string. If there is a type mismatch. Please fix either the onSubmit function or the value passed here should be a string.

Reply:
Ig the types have been mismatched in almost all migrated interactions. Like this one
https://github.com/oppia/oppia/blob/1502caf43bd62019d94478391a1e0b387a3a73c1/extensions/interactions/FractionInput/directives/oppia-interactive-fraction-input.component.ts#L184-L187
and this one
https://github.com/oppia/oppia/blob/1502caf43bd62019d94478391a1e0b387a3a73c1/extensions/interactions/NumericInput/directives/oppia-interactive-numeric-input.component.ts#L71-L73
There are lot of other instances as well. This should be dealt with in a separate PR, otherwise there'll be so many changes here.

Reply:
Please add a TODO.